File: technology/java-spring-boot/src/test/java/org/acme/schooltimetabling/rest/TimetableFastAssertTest.java
package org.acme.schooltimetabling.rest;

import static org.awaitility.Awaitility.await;

import java.time.Duration;

import ai.timefold.solver.core.api.solver.SolverStatus;

import org.acme.schooltimetabling.domain.Timetable;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.condition.DisabledInNativeImage;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.web.server.LocalServerPort;
import org.springframework.test.web.reactive.server.WebTestClient;

import com.fasterxml.jackson.databind.JsonNode;

@SpringBootTest(properties = {
        "timefold.solver.environment-mode=FAST_ASSERT",
        "timefold.solver.termination.spent-limit=1s",
        "logging.level.ai.timefold.solver=ERROR" },
        webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@Tag("slowly")
@DisabledInNativeImage
class TimetableFastAssertTest {

    @LocalServerPort
    private int port;

    @Test
    void solve() {
        WebTestClient client = WebTestClient.bindToServer()
                .baseUrl("http://localhost:" + port)
                .build();

        Timetable testTimetable = client.get()
                .uri("/demo-data/SMALL")
                .exchange()
                .expectBody(Timetable.class)
                .returnResult()
                .getResponseBody();

        String jobId = client.post()
                .uri("/timetables")
                .bodyValue(testTimetable)
                .exchange()
                .expectBody(String.class)
                .returnResult()
                .getResponseBody();

        await()
                .atMost(Duration.ofMinutes(1))
                .pollInterval(Duration.ofMillis(500L))
                .until(() -> SolverStatus.NOT_SOLVING.name().equals(
                        client.get()
                                .uri("/timetables/" + jobId + "/status")
                                .exchange()
                                .expectBody(JsonNode.class)
                                .returnResult()
                                .getResponseBody()
                                .get("solverStatus")
                                .asText()));

        client.get()
                .uri("/timetables/" + jobId)
                .exchange()
                .expectBody()
                .jsonPath("score").isNotEmpty();
    }

}

File: technology/kotlin-quarkus/src/test/kotlin/org/acme/kotlin/schooltimetabling/rest/TimetableFastAssertTest.kt
package org.acme.kotlin.schooltimetabling.rest

import ai.timefold.solver.core.api.solver.SolverStatus
import io.quarkus.test.junit.QuarkusTest
import io.quarkus.test.junit.QuarkusTestProfile
import io.quarkus.test.junit.TestProfile
import io.restassured.RestAssured.get
import io.restassured.RestAssured.given
import io.restassured.http.ContentType
import org.acme.kotlin.schooltimetabling.domain.Timetable
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import java.time.Duration

@QuarkusTest
@TestProfile(TimetableFastAssertTest.FullAssertProfile::class)
@Tag("slowly")
class TimetableFastAssertTest {

    @Test
    fun solve() {
        val testTimetable: Timetable = given()
            .`when`()["/demo-data/SMALL"]
            .then()
            .statusCode(200)
            .extract()
            .`as`(Timetable::class.java)

        val jobId: String = given()
            .contentType(ContentType.JSON)
            .body(testTimetable)
            .expect().contentType(ContentType.TEXT)
            .`when`().post("/timetables")
            .then()
            .statusCode(200)
            .extract()
            .asString()

        await()
            .atMost(Duration.ofMinutes(1))
            .pollInterval(Duration.ofMillis(500L))
            .until {
                SolverStatus.NOT_SOLVING.name ==
                        get("/timetables/$jobId/status")
                            .jsonPath().get("solverStatus")
            }
        val solution: Timetable =
            get("/timetables/$jobId").then().extract().`as`<Timetable>(
                Timetable::class.java
            )
        assertTrue(solution.score?.isFeasible!!)
    }

    class FullAssertProfile : QuarkusTestProfile {
        override fun getConfigOverrides(): Map<String, String> {
            return java.util.Map.of(
                "quarkus.timefold.solver.environment-mode", "FAST_ASSERT",
                "quarkus.timefold.solver.termination.best-score-limit", "",
                "quarkus.timefold.solver.termination.spent-limit", "30s"
            )
        }
    }
}
